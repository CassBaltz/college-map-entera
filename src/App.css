.main {
  padding: 24px;
  display: flex;
  flex-direction: column;
  height: 100%;
}

#map {
  margin-top: 16px;
  border-radius: 8px;
  width: 100%;
  flex-grow: 2;
}

.schoolSearchForm {
  display: flex;
  height: 48px;
  align-items: stretch;
}

.schoolSearchForm input {
  width: 50%;
  border-radius: 4px;
  margin-right: 16px;
  padding: 8px;
  border-color: #ced4da;
  font-size: 18px;
}

.schoolSearchForm button {
  outline: none;
  box-shadow: none;
  background-color: #339af0;
  border-color: #1c7ed6;
  border-radius: 4px;
  color: white;
  padding: 8px;
  box-shadow: none;
  font-size: 18px;
}

.schoolSearchForm button:disabled {
  background-color: #4dabf7;
}

.errorsWrapper {
  margin: 16px 0 0;
  padding: 16px;
  background-color: #fff5f5;
  color: #fa5252;
  list-style-type: none;
  border-radius: 4px;
}

.errorsWrapper li {
  margin: 8px 0;
}
