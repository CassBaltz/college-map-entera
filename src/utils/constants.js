// Normally these would go into a secrets or env config file

export const COLLEGE_SEARCH_BASE_URL = "https://api.data.gov";
export const COLLEGE_SEARCH_PATH = "/ed/collegescorecard/v1/schools.json";
export const COLLEGE_SEARCH_API_KEY =
  "KalKbXfAbfU304B29CpuO7XvfLR3JJztfNX5VLev";
export const GOOGLE_MAPS_API_KEY = "AIzaSyDR07l6fNwbdlmSp0kjrhzeF4Er8dOefnQ";
