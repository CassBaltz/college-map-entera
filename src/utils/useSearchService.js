import { useEffect, useState } from "react";
import { Loader } from "@googlemaps/js-api-loader";
import {
  COLLEGE_SEARCH_BASE_URL,
  COLLEGE_SEARCH_PATH,
  COLLEGE_SEARCH_API_KEY,
  GOOGLE_MAPS_API_KEY,
} from "./constants";
import axios from "axios";

const axiosInstance = axios.create({
  baseURL: COLLEGE_SEARCH_BASE_URL,
});

let googleMapInstance = null;

const buildSchool = ({ latest, location }, map) => ({
  name: latest.school.name,
  location,
  marker: new global.google.maps.Marker({
    title: latest.school.name,
    position: {
      lat: location.lat,
      lng: location.lon,
    },
    map,
  }),
});

const useSearchService = () => {
  const [errors, setErrors] = useState([]);
  const [schools, setSchools] = useState([]);
  const [searching, setSearching] = useState(false);

  useEffect(() => {
    new Loader({
      apiKey: GOOGLE_MAPS_API_KEY,
    })
      .load()
      .then(() => {
        const map = new global.google.maps.Map(document.getElementById("map"), {
          // Best guess for middle of the United States
          center: { lat: 37, lng: -98 },
          zoom: 4.5,
        });

        googleMapInstance = map;
      });
  }, []);

  const startSearch = () => {
    setSearching(true);
    setErrors([]);
    schools.forEach((school) => {
      school.marker.setMap(null);
    });
    setSchools([]);
  };

  const searchSchool = (schoolName) => {
    startSearch();
    axiosInstance
      .get(COLLEGE_SEARCH_PATH, {
        params: {
          api_key: COLLEGE_SEARCH_API_KEY,
          school_search: schoolName,
        },
      })
      .then(({ data }) => {
        setSearching(false);
        const { results } = data;

        if (results.length === 0) {
          setErrors([
            { message: `Could not find any matches for ${schoolName}` },
          ]);
        }

        const updatedSchools = results.map((school) =>
          buildSchool(school, googleMapInstance),
        );

        setSchools(updatedSchools);
      })
      .catch((error) => {
        setSearching(false);
        setErrors([
          {
            message: "Unknown error occurred searching, please try again later",
          },
        ]);
      });
  };

  return {
    errors,
    searching,
    searchSchool,
  };
};

export { useSearchService, buildSchool };
