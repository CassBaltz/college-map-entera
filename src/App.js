import "./App.css";
import SearchForm from "./components/search-form";
import Errors from "./components/errors";
import { useSearchService } from "./utils/useSearchService";

const generateHandleSubmit = (searchSchool, searching) => (e) => {
  e.preventDefault();
  if (searching) return;

  const formData = new FormData(e.target);
  const schoolName = formData.get("school-name");

  searchSchool(schoolName);
};

const App = () => {
  const { searchSchool, errors, searching } = useSearchService();

  return (
    <div className="main">
      <h1>College Search</h1>
      <SearchForm
        searching={searching}
        handleSubmit={generateHandleSubmit(searchSchool, searching)}
      />
      <Errors errors={errors} />
      <div id="map" />
    </div>
  );
};

export default App;
