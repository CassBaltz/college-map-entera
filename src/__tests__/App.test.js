import { shallow } from "enzyme";
import App, { generateHandleSubmit } from "../App";
import SearchForm from "../components/search-form";
import Errors from "../components/errors";
import { useSearchService } from "../utils/useSearchService";

jest.mock("../utils/useSearchService", () => ({
  useSearchService: jest.fn(),
}));

describe("App", () => {
  it("should render correctly - initial state", () => {
    const searchSchool = jest.fn();
    const errors = [];
    const searching = false;

    useSearchService.mockImplementation(() => ({
      searchSchool,
      errors,
      searching,
    }));

    const app = shallow(
      <App searchSchool={searchSchool} errors={errors} searching={searching} />,
    );

    const errorsComp = app.find(Errors);
    const searchFormComp = app.find(SearchForm);
    const map = app.find("#map");

    expect(app.getElement()).toMatchSnapshot();

    expect(map).toHaveLength(1);

    expect(searchFormComp).toHaveLength(1);
    expect(searchFormComp.prop("searching")).toBe(searching);

    expect(errorsComp).toHaveLength(1);
    expect(errorsComp.prop("errors")).toBe(errors);
  });
});

describe("generateHandleSubmit", () => {
  it("should not do anything if search is in progress", () => {
    const searchSchool = jest.fn();
    const searching = true;
    const preventDefault = jest.fn();

    const event = {
      preventDefault,
    };

    generateHandleSubmit(searchSchool, searching)(event);

    expect(preventDefault).toHaveBeenCalled();
    expect(searchSchool).not.toHaveBeenCalled();
  });

  it("should parse the data from the target element and call the searchSchool function with the input value", () => {
    const searchSchool = jest.fn();
    const searching = false;
    const preventDefault = jest.fn();

    const mockForm = document.createElement("form");
    mockForm.setAttribute("name", "school-search");

    const mockInput = document.createElement("input");
    mockInput.setAttribute("name", "school-name");
    mockInput.setAttribute("value", "Davidson");

    mockForm.appendChild(mockInput);

    const event = {
      preventDefault,
      target: mockForm,
    };

    generateHandleSubmit(searchSchool, searching)(event);
    expect(searchSchool).toHaveBeenCalledWith("Davidson");
  });
});
