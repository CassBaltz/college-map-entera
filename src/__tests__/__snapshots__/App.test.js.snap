// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly - initial state 1`] = `
<div
  className="main"
>
  <h1>
    College Search
  </h1>
  <SearchForm
    handleSubmit={[Function]}
    searching={false}
  />
  <Errors
    errors={Array []}
  />
  <div
    id="map"
  />
</div>
`;
